src/neuroglancer/perspective_view/panel.ts

```diff
@@ -225,6 +226,39 @@ export class PerspectivePanel extends RenderedDataPanel {
       });
     });
 
+    registerActionListener(element, 'translate-annoation-via-mouse-drag', (e: ActionEvent<MouseEvent>) => {
+      const {mouseState} = this.viewer;
+      const selectedAnnotationId = mouseState.pickedAnnotationId;
+      const annotationLayer = mouseState.pickedAnnotationLayer;
+
+      let voxelSize = this.viewer.navigationState.voxelSize
+      if (typeof(annotationLayer) != 'undefined'){
+        if (typeof(selectedAnnotationId) != 'undefined'){
+          
+          let annotationRef = annotationLayer.source.getReference(selectedAnnotationId)!;
+          let ann = <Annotation>annotationRef.value
+          
+          const handler = getAnnotationTypeHandler(ann.type)
+          let point = handler.getClosestPoint(ann, voxelSize, mouseState.position)
+
+          if (mouseState.updateUnconditionally()) {
+            startRelativeMouseDrag(e.detail, (_event, deltaX, deltaY) => {
+              const temp = tempVec3;
+              const {projectionMat} = this;
+              const {width, height} = this;
+      
+              voxelSize.spatialFromVoxel(point,point)
+              vec3.transformMat4(temp, point, projectionMat);
```

Comment:
The annotations are intended to be immutable, so instead of modifying point here, instead create a new annotation with the point modified, and call annotationLayer.source.update(annotationRef, newAnnotation);

Finally, once the drag ends, call annotationLayer.source.commit(annotationRef);  and annotationRef.dispose(); to avoid leaking a reference.

Unfortunately, startRelativeMouseDrag doesn't currently provide a way to specifying a callback to be invoked when the drag ends (needed into order to call commit), but that would be easy to add as an optional additional argument.

For the in-memory annotations, just updating in place and calling source.changed.dispatch works, as you've noticed, but for annotations stored on a server, the update/commit calls are needed.

Reply:
Thanks for the feedback.  Seems all straightforward andI have most done already, except I'm a bit confused about how to move forward with the pickedOffset.  

We'd have to translate each of these pickedOffsets to a desired behavior for each annotation.. i.e. clicking on the line should translate both endpoints for the Line annotation, but clicking on a endpoint should only move one, so maybe we'd have to have an interface for each annotation type which would give a point for an annotation and its pickedOffset, so the handler put into the Drag callback would move that point, and then another interface which would update the annotation's internal representation given a new point and the same pickedOffset?  

something like this (for sliceView?) (with getRepresentativePoint and updateViaRepresentativePoint to be implemented by each AnnotationType).
``` typescript
          let annotationRef = annotationLayer.source.getReference(selectedAnnotationId)!;
          let ann = <Annotation>annotationRef.value;
          let newAnnotation = {...ann};
          annotationLayer.source.update(annotationRef, newAnnotation);
          const handler = getAnnotationTypeRenderHandler(newAnnotation.type);
          let repPoint = handler.getRepresentativePoint(ann, mouseState.pickedOffset);
          if (mouseState.updateUnconditionally()) {
            startRelativeMouseDrag(e.detail, (_event, deltaX, deltaY) => {
              
              vec3.transformMat4(repPoint, repPoint, this.sliceView.dataToViewport);
              vec3.set(repPoint, repPoint[0]-deltaX, repPoint[1]-deltaY, point[2]);
              vec3.transformMat4(repPoint, repPoint, this.sliceView.viewportToData);
              handler.updateViaRepresentativePoint(ann, mouseState.pickedOffset, repPoint)
              annotationLayer.source.changed.dispatch();
            },
            (_event) => {
              annotationRef.value = newAnnotation;
              annotationLayer.source.commit(annotationRef);
              annotationRef.dispose();
            });
         }
```

Another issue I encountered when exploring this... 
Seems like for a Line it returns 0  for the line, 1 for pointA and 2 for pointB, but for BoundingBox it only returns 0 (the FULL_OBJECT_PICK_OFFSET) no matter where I click on the box in the sliceView (though I get different indices for each line of the box in perspectiveView).  I don't think the corner's are selectable in any of the views.. probably because they don't have shaders... but that's maybe not a problem. Do you know why and how I might fix that?

Reply:
oh and this assumes of course that I have the new optional finishMoveHandler option on startRelativeMouseDrag

```typescript
export function startRelativeMouseDrag(initialEvent: MouseEvent,
                                       handler: RelativeDragHandler,
                                       finishDragHandler?: FinishDragHandler ) {
  let {document} = initialEvent.view;
  let prevScreenX = initialEvent.screenX, prevScreenY = initialEvent.screenY;
  let mouseMoveHandler = (e: MouseEvent) => {
    let deltaX = prevScreenX - e.screenX;
    let deltaY = prevScreenY - e.screenY;
    prevScreenX = e.screenX;
    prevScreenY = e.screenY;
    handler(e, deltaX, deltaY);
  };
  let button = initialEvent.button;
  let mouseUpHandler = (e: MouseEvent) => {
    if (e.button === button) {
      document.removeEventListener('mousemove', mouseMoveHandler, true);
      document.removeEventListener('mouseup', mouseUpHandler, false);
      if (typeof(finishDragHandler) != 'undefined'){
        finishDragHandler(e);
      }
    }
  };
  document.addEventListener('mousemove', mouseMoveHandler, true);
  document.addEventListener('mouseup', mouseUpHandler, false);
}
```